src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageSource.java

```diff
@@ -0,0 +1,355 @@
+		this.ackCallbackFactory = ackCallbackFactory;
+		this.topics = topics;
+	}
+
+	protected String getGroupId() {
+		return this.groupId;
+	}
+
+	/**
+	 * Set the group.id property for the consumer.
+	 * @param groupId the group id.
+	 */
+	public void setGroupId(String groupId) {
+		this.groupId = groupId;
+	}
+
+	protected String getClientId() {
+		return this.clientId;
+	}
+
+	/**
+	 * Set the client.id property for the consumer.
+	 * @param clientId the client id.
+	 */
+	public void setClientId(String clientId) {
+		this.clientId = clientId;
+	}
+
+	protected long getPollTimeout() {
+		return this.pollTimeout;
+	}
+
+	/**
+	 * Set the pollTimeout for the poll() operations; default 50ms.
+	 * @param pollTimeout the poll timeout.
+	 */
+	public void setPollTimeout(long pollTimeout) {
+		this.pollTimeout = pollTimeout;
+	}
+
+	protected RecordMessageConverter getMessageConverter() {
+		return this.messageConverter;
+	}
+
+	/**
+	 * Set the messsage converter to replace the default
+	 * {@link MessagingMessageConverter}.
+	 * @param messageConverter the converter.
+	 */
+	public void setMessageConverter(RecordMessageConverter messageConverter) {
+		this.messageConverter = messageConverter;
+	}
+
+	protected Type getPayloadType() {
+		return this.payloadType;
+	}
+
+	/**
+	 * Set the payload type.
+	 * Only applies if a type-aware message converter is provided.
+	 * @param payloadType the type to convert to.
+	 */
+	public void setPayloadType(Type payloadType) {
+		this.payloadType = payloadType;
+	}
+
+	@Override
+	public String getComponentType() {
+		return "kafka:message-source";
+	}
+
+	@Override
+	protected synchronized AbstractIntegrationMessageBuilder<Object> doReceive() {
+		if (this.consumer == null) {
+			createConsumer();
+		}
+		ConsumerRecord<K, V> record;
+		if (this.recordIterator != null && this.recordIterator.hasNext()) {
+			record = this.recordIterator.next();
+		}
+		else {
+			synchronized (this.consumer) {
+				try {
+					Set<TopicPartition> paused = this.consumer.paused();
+					if (paused.size() > 0) {
+						this.consumer.resume(paused);
+					}
+					ConsumerRecords<K, V> records = this.consumer.poll(this.pollTimeout);
+					this.consumer.pause(this.partitions);
+					if (records == null || records.count() == 0) {
+						this.recordIterator = null;
+						return null;
+					}
+					this.recordIterator = records.iterator();
+					record = this.recordIterator.next();
+				}
+				catch (WakeupException e) {
+					return null;
+				}
+			}
+		}
+		@SuppressWarnings("unchecked")
+		Message<Object> message = (Message<Object>) this.messageConverter.toMessage(record, null, null,
+				this.payloadType);
+		AcknowledgmentCallback ackCallback = this.ackCallbackFactory
+				.createCallback(new KafkaAckInfo<K, V>(this.consumer, record, this::resetIterator));
+		return getMessageBuilderFactory().fromMessage(message)
+				.setHeader(IntegrationMessageHeaderAccessor.ACKNOWLEDGMENT_CALLBACK, ackCallback);
```

Comment:
I think we need to do a check for the `if (message.getHeaders() instanceof KafkaMessageHeaders) {` here as well.
For optimization. Similar what you've done recently in the `KafkaMessageDrivenChannelAdapter` for the `IntegrationMessageHeaderAccessor.DELIVERY_ATTEMPT` header

Reply:
Missed this one; we do tomorrow